### ISSUE TITLE
webhook, missing basket

### ISSUE THREAD

**Issue by j2l** [courtneypattison/thegrouch#3]
Hello,
It looks like "Basket" is missing:
```
Internal Server Error: /stripe/webhook/
Traceback (most recent call last):
  File "/home/pm/Documents/github/django_tests/thegrouch-stripe-restapi/.venv/lib/python3.10/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
  File "/home/pm/Documents/github/django_tests/thegrouch-stripe-restapi/.venv/lib/python3.10/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/home/pm/Documents/github/django_tests/thegrouch-stripe-restapi/.venv/lib/python3.10/site-packages/django/views/generic/base.py", line 104, in view
    return self.dispatch(request, *args, **kwargs)
  File "/home/pm/Documents/github/django_tests/thegrouch-stripe-restapi/.venv/lib/python3.10/site-packages/django/utils/decorators.py", line 46, in _wrapper
    return bound_method(*args, **kwargs)
  File "/home/pm/Documents/github/django_tests/thegrouch-stripe-restapi/.venv/lib/python3.10/site-packages/django/views/decorators/csrf.py", line 56, in wrapper_view
    return view_func(*args, **kwargs)
  File "/home/pm/Documents/github/django_tests/thegrouch-stripe-restapi/.venv/lib/python3.10/site-packages/oscar/apps/checkout/session.py", line 68, in dispatch
    return super().dispatch(request, *args, **kwargs)
  File "/home/pm/Documents/github/django_tests/thegrouch-stripe-restapi/.venv/lib/python3.10/site-packages/django/views/generic/base.py", line 143, in dispatch
    return handler(request, *args, **kwargs)
  File "/home/pm/Documents/github/django_tests/thegrouch-stripe-restapi/thegrouch/thegrouch/apps/stripe/views.py", line 35, in post
    basket = Basket.objects.get(pk=self.checkout_session.get_submitted_basket_id())
  File "/home/pm/Documents/github/django_tests/thegrouch-stripe-restapi/.venv/lib/python3.10/site-packages/django/db/models/manager.py", line 87, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/home/pm/Documents/github/django_tests/thegrouch-stripe-restapi/.venv/lib/python3.10/site-packages/django/db/models/query.py", line 637, in get
    raise self.model.DoesNotExist(
oscar.apps.basket.models.Basket.DoesNotExist: Basket matching query does not exist
```
Also in the Stripe Received Events, there's no sign of Basket.

I'm using the latest API version (2023-10-16), Django 4.2.9, Django-Oscar 3.2.2, stripe 8.0.0
Which versions work for you? 
Any idea what is wrong in my case?

Note: If I were you, I would remove/replace the db.sqlite3, it includes the test order you made on September 8, it has a real address in Kitchener.

**Comment by courtneypattison** [comment]
Thank you so much for letting me know about the db address! Super sloppy of me.

I don't think I can be much help. This project was for a take home interview so I rushed through it and I have no previous experience with Django or Stripe. I put it in GitHub in the hopes that it might help someone else since the examples I found used Stripe's old API. It seems like it's causing you more trouble than good though!

It works for me with Django 3.2.21, Django-Oscar 3.2 and Stripe 6.5.0.

I'm probably way off, but maybe make sure your events are getting correctly [forwarded to your webhook](https://dashboard.stripe.com/test/webhooks/create?endpoint_location=local).

**Comment by j2l** [comment]
Thank you very much for your honest review of your process and sharing this project. It actually does help me a lot to learn Django-Oscar. I found too that other DO projects are using the old stripe API, or decommissioned libraries (paypal), which is weird.

Does the webhook currently fully work for you? No error about Basket?
If so, could you please do a `pip freeze > thegrouch-requirements.txt` and post it here?

**Comment by j2l** [comment]
Since I'm learning, and you seem to be a wiz with python (you've done it without knowing Django and Oscar, I couldn't have done this project from scratch with my knowledge), may I ask how you approached it? 
From the stripe API code (for flask)? From another project? From Oscar libraries?
What were the steps to get to this result?
I'm asking all this because I'm not sure I found the right way to approach these frameworks. Information is scattered or old, even doc is outdated.

**Comment by j2l** [comment]
Hello and sorry to bother you, does the webhook currently fully work for you? No error about Basket?

**Comment by j2l** [comment]
Thank you so much @courtneypattison !! This is so detailed and helpful.
I will study all this closely and let you know. I'm pretty sure I missed a tiny thing.

Good luck with all you have to do!

**Comment by j2l** [comment]
I updated, package by package (downgraded from Django 4.2.9 to 3.2.20, downgraded oscar, ...) flushed the db, made migrations, migrated, runserver,  ... no luck so far but a couple of things:

It looks like there are 2 routes, depending whether you run (in a different bash, while server is running) `stripe listen --forward-to 127.0.0.1:8000/stripe/webhook/` or not.


If you do, like me (to write events in stripe local webhook log), you should get the call to the webhook, followed by the basket error:
```
[07/Feb/2024 11:49:29] "POST /checkout/preview/ HTTP/1.1" 302 0
[07/Feb/2024 11:49:48] "POST /stripe/webhook/ HTTP/1.1" 200 0
...
oscar.apps.basket.models.Basket.DoesNotExist: Basket matching query does not exist.
```

If you don't run it, it is like your log:
```
[07/Feb/2024 11:53:36] "POST /checkout/preview/ HTTP/1.1" 302 0
[07/Feb/2024 11:53:58] "GET /checkout/thank-you/ HTTP/1.1" 302 0
[07/Feb/2024 11:53:58] "GET /catalogue/ HTTP/1.1" 200 10081
```
No error ... but no order passed (at least for me).

Can you check if any recent order was created for you in the dashboard (http://localhost:8000/dashboard/orders/) or django admin (http://localhost:8000/admin/order/order/)?

If not, there's something bypassing the webhook to go straight to the thank-you page (that doesn't exist, 302).

**Comment by courtneypattison** [comment]
@j2l You're right! I forgot about setting up the stripe listener. I'm getting the same error message. Looks like I never had it configured correctly. I have no time to get it working. Thank you for your help and sorry for the trouble.

**Comment by j2l** [comment]
OK, I understand. I don't want to bother you. When you have 5min:
May I just ask how did you get the order in the first place?
